### ISSUE TITLE
[Routing] Url Generator and slashes in url parameters

### ISSUE THREAD

**Issue by Imunhatep** [symfony/symfony#13017]
Currently symfony does not allow to generate urls with encoded slash in slug. Ex:
```yml
_some_route:
    path: /create/{folder}/zip
```
If `folder` parameter value is: `/home/user`, either I have to allow `/` in url parameter or do parameter encoding before passing it to the UrlGenerator. 
In the first case we get raw path: `/create//home/user/zip`
In the second case, UrlGeneretor will double encode the parameter: `/create/%252Fhome%252Fuser/zip`.
But neither give desired output: `/create/%2Fhome%2Fuser/zip`

I saw remark in the UrlGenerator code to problems with JBOSS and Apache for url with encoded slashes (http://stackoverflow.com/questions/4069002/http-400-if-2f-part-of-get-url-in-jboss), also it looks more like a bug to me, nevertheless I think it should be handled by programmers that work with that systems, cause encoded slash is fully valid in url. 

Also current approach forces to use double decoding on servers with proper handling of urls, or overriding url generator class in symfony.

I do understand that my proposal will break back-compatibility, but I think this behavior should be discussed at least for future versions.

**Comment by nacmartin** [comment]
Even when `%2F` is, strictly speaking, allowed in URLs, if it is disallowed in Apache and JBoss for security reasons, it makes it in practice disallowed, IMO. Apache is not just another marginal web server and Symfony cannot just say "let their users deal with the problem".

Anyways, could this be solved by adding a method `setDecodedChars` so you could override [the default pairs of decodedChars](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Routing/Generator/UrlGenerator.php#L60)?

**Comment by Imunhatep** [comment]
That is not fix I was hoping for..  but yes, that will help, thought not for current implementation of UrlGenerator class, it will incorrectly encode url because of code at line 192.

IMHO: There're way too many other web servers (nginx, tomcat, netty, jetty, e.t.c.), if someone introduced something stupid, it doesn't mean that others should support it. It's a bit strange to have advantage for 2 servers, and break compatibility with others. A specially with URI RFC.

P.S. That apache and jboss behavior is configurable.

**Comment by teohhanhui** [comment]
Double URL encoding could be a security risk: https://www.owasp.org/index.php/Double_Encoding

**Comment by teohhanhui** [comment]
Perhaps instead of

https://github.com/symfony/symfony/blob/b85ab60b70f0dcff62ac8438e91e7a93c44823da/src/Symfony/Component/Routing/Generator/UrlGenerator.php#L188

we can do the `rawurlencode` before we prepend to `$url`, and for each of them we can detect whether they're already URL-encoded (is that safe?) or best if the user can set an `already_encoded` flag for each parameter...

The `strtr` seems safe to be done only once at the current line.

**Comment by HeahDude** [comment]
Solutions proposed by @Tobion should be the way to go.

Let's close this old issue as "won't fix", since it's not a good practice to resolve parameters containing slashes.

Thanks.

**Comment by MarioLipinski** [comment]
I strongly disagree. A framework such as Synfomy should be able to handle and generate all kind of standard compliant URIs. If, for any reason, certain properly encoded characters should be forbidden in the path, this should be done by some **configurable** validation component. Given that this is not possible with the current UrlGenerator is a strong indication to rewrite the UrlGenerator.

**Comment by Imunhatep** [comment]
2016.. The most popular php **web** framework, is not compatible with **uri** rfc (rfc3986). 
RFC, Carl!

**Comment by lazka** [comment]
What's the status of this? Are there any workarounds that don't require changes to the URL schema?
